### PR TITLE
Feature Server

### PR DESCRIPTION
**Pull request by jammiess**
## Description

Adds a small gRPC server that can be used for adding features to Architus in a closed or open beta state.

Adds the ability to selectively add or remove features from guilds. This mainly adds the functionality of letting guilds opt-in to beta features that may or may not be fully implemented yet.

## Checklist

- [ ] [Linked PR to corresponding issue in this repo](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue)
- [x] Manually tested locally
- [ ] Additional issue created in [docs repo](https://github.com/architus/docs.archit.us/issues) (if necessary)

### CODE REVIEW

File: feature-gate/proto/feature-gate.proto [hunk 1]
```diff
@@ -0,0 +1,58 @@
+syntax = "proto3";
```
Comment (johnyburd) [inline feature-gate/proto/feature-gate.proto:1]: can we not use just the copy of this file in `/lib/ipc`?

File: feature-gate/grpc/featuregate.rs [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+pub struct FeatureName {
+    #[prost(string, tag = "1")]
+    pub name: std::string::String,
+}
+#[derive(Clone, PartialEq, ::prost::Message)]
+pub struct FeatureAddition {
+    #[prost(string, tag = "1")]
+    pub feature_name: std::string::String,
+    #[prost(sfixed64, tag = "2")]
+    pub guild_id: i64,
```
Comment (johnyburd) [inline feature-gate/grpc/featuregate.rs:18]: I believe these should be `u64`s technically, even though we use the signed data type in postgres 😕

File: db/architus-seed.sql [hunk 3]
```diff
@@ -149,6 +149,34 @@ CREATE TABLE public.tb_users (
 
 ALTER TABLE public.tb_users OWNER TO autbot;
 
+--
+-- Name: tb_features; Type: TABLE; Schema: public; Owner: autbot
```
Comment (johnyburd) [inline db/architus-seed.sql:153]: nice name :)

File: feature-gate/build.rs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+        .status()?;
+
+    tonic_build::configure()
+        .build_client(false)
+        .build_server(true)
+        .format(true)
+        .out_dir("./grpc")
+        .compile(&["feature-gate.proto"], &["./proto"])?;
+
+    // Clippy is dumb and doesn't have an easy way to exclude autogenerated code.
```
Comment (johnyburd) [inline feature-gate/build.rs:16]: lol gross

File: feature-gate/src/bin/feature-gate.rs [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+
+        let feature = request.into_inner();
+        let result = insert_feature(&conn, &feature.name, feature.open);
+
+        match result {
+            Ok(_) => {
+                info!(
+                    "Added feature {} which is {}",
+                    feature.name,
+                    if feature.open { "open" } else { "closed" }
```
Comment (johnyburd) [inline feature-gate/src/bin/feature-gate.rs:67]: nice ternaries, rust

File: feature-gate/proto/feature-gate.proto (hunk 1 above)
Comment (jammiess) [reply feature-gate/proto/feature-gate.proto:1]: Lol. That actually is what happens currently. It just copies that into a local file for building and I accidentally pushed that.

**Pull request merged**